### ISSUE TITLE
Crash loop when specifying port as number instead of string in ServiceMonitor

### ISSUE THREAD

**Issue by olvesh** [coreos/prometheus-operator#892]
**What did you do?**
Tested if a ServiceMonitor could handle port numbers

**What did you expect to see?**
A prometheus scraping services using a standard port number, not by name. 

**What did you see instead? Under which circumstances?**
This was only a test. I am not saying that promehteus-operator should handle port numbers in addition to port names in the ServiceMonitor (although that would be cool) but I did not expect the operator to enter a crash loop.

Now coreos-operator is in a crashloop with the message:
`ts=2018-01-15T09:35:38Z caller=main.go:147 msg="Unhandled error received. Exiting..." err="creating CRDs failed: timed out waiting for Custom Resoruce: json: cannot unmarshal number into Go struct field Endpoint.port of type string"`

**Environment**
Kubernetes 1.8.5

* Kubernetes version information:

```$ kubectl version
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.2", GitCommit:"bdaeafa71f6c7c04636251031f93464384d54963", GitTreeState:"clean", BuildDate:"2017-10-24T19:48:57Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.5-gke.0", GitCommit:"2c2a807131fa8708abc92f3513fe167126c8cce5", GitTreeState:"clean", BuildDate:"2017-12-19T20:05:45Z", GoVersion:"go1.8.3b4", Compiler:"gc", Platform:"linux/amd64"}````

* Kubernetes cluster kind: 
GKE

* Manifests:
Helm excerpt:
```
apiVersion: {{ template "prometheus-operator.apiVersion" . }}
kind: ServiceMonitor
metadata:
  labels:
    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
    component: "{{ .Release.Name }}"
    heritage: "{{ .Release.Service }}"
    release: "{{ .Release.Name }}"
    prometheus: {{ .Values.prometheusName }}
  name: {{ template  "dt-monitor-cloud.fullname" . }}
spec:
  jobLabel: {{ template  "dt-monitor-cloud.fullname" . }}
  selector:
    matchExpressions:
      - {key: app, operator: Exists}
  namespaceSelector:
    matchNames:
      - {{ .Values.targetNamespace | quote }}
  endpoints:
  - port:  9102
    interval: 15s
```

* Prometheus Operator Logs:

```
ts=2018-01-15T09:35:32Z caller=operator.go:257 component=prometheusoperator msg="connection established" cluster-version=v1.8.5-gke.0
ts=2018-01-15T09:35:32Z caller=operator.go:156 component=alertmanageroperator msg="connection established" cluster-version=v1.8.5-gke.0
ts=2018-01-15T09:35:32Z caller=operator.go:170 component=alertmanageroperator msg="CRD API endpoints ready"
ts=2018-01-15T09:35:32Z caller=operator.go:303 component=alertmanageroperator msg="Alertmanager added" key=monitoring/dt-prometheus-infra
ts=2018-01-15T09:35:32Z caller=operator.go:379 component=alertmanageroperator msg="sync alertmanager" key=monitoring/dt-prometheus-infra
ts=2018-01-15T09:35:32Z caller=operator.go:323 component=alertmanageroperator msg="Alertmanager updated" key=monitoring/dt-prometheus-infra
ts=2018-01-15T09:35:32Z caller=operator.go:1035 component=prometheusoperator msg="CRD created" crd=Prometheus
ts=2018-01-15T09:35:32Z caller=operator.go:1035 component=prometheusoperator msg="CRD created" crd=ServiceMonitor
ts=2018-01-15T09:35:32Z caller=operator.go:379 component=alertmanageroperator msg="sync alertmanager" key=monitoring/dt-prometheus-infra
ts=2018-01-15T09:35:38Z caller=main.go:147 msg="Unhandled error received. Exiting..." err="creating CRDs failed: timed out waiting for Custom Resoruce: json: cannot unmarshal number into Go struct field Endpoint.port of type string"```

**Comment by brancz** [comment]
This is expected. Unfortunately we have not yet gotten around to implementing CRD validations, but the problem here is that port is supposed to be the name of a port. If you want to specify an explicit number, use `targetPort`. Here are the [API Docs](https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#endpoint).

**Comment by olvesh** [comment]
Sure, I understand that it is wrong input in the service monitor, but should that force the operator to enter a crash loop?

Also when the operator was in a crash loop, I did a deploy of CRD Alertmanager and Prometheus. Helm reported success, but the CRD instances was nowhere to be seen afterwards. I would think that the system would balance itself after I fixed the failing port number..

Now helm thinks everything is ok, but the alertmanager and prometheus definition is nowhere to be seen. I did not reinsrtall operator along with CRDs, just fixed the erroneous ServiceMonitor.
